presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -1219,6 +1246,30 @@ private InnerJoinPushDownResult processInnerJoin(RowExpression inheritedPredicat
                 }
             }
 
+            if (shouldGenerateDomainFilters(session)) {
+                // Extract domains for each of the variables from the inherited predicate
+                // and translate them back to predicates that can be added to the sources
+                // These prove to be helpful to extract predicates on columns that cannot be converted to CNF cleanly
+                // E.g. for [(Left=4 AND Right=20) or (Left=5 AND Right=21) or (Left=6 AND Right=22)]
+                // We would extract the TupleDomain = [ Left IN (4,5,6), Right IN (20,21,22)], then convert them into predicates on 'Left' & 'Right'
```

Comment:
I suppose this only handles literal value comparisons and not expression that cannot be constant folded.
Can you also add negative tests that contain expressions, e.g. 
```
Left=Right+1 and Right=20 or ...
```

Reply:
Yes, equality inference only works with plain conjunctions. I added a test `testComplexDisjunctExpression` to highlight this